### PR TITLE
[BUG FIX] Fix convex collision detection on geom pairs of very different sizes.

### LINKED ISSUE #3314: [Bug]: mpr_refine_portal has no iteration cap; a non-converging portal spins a GPU lane forever (standalone reproduction attached)
## Bug Description

`mpr_refine_portal` (`genesis/engine/solvers/rigid/collider/mpr.py:263-310`, the `while True` at line 280) has no iteration cap. Its sibling `mpr_find_penetration` is capped by `CCD_ITERATIONS`, and libccd caps both loops. On near-degenerate geometry the portal keeps expanding by rounding-sized steps without encapsulating the origin or reaching the tolerance: one lane spins forever, the kernel never returns, and the host blocks at its next device synchronisation. No NaN or blown-up state is involved, and the point at which it happens is deterministic given the trajectory.

Observed on a dual-quadruped push (two Go1 URDFs, a box, a plane), 256–512 envs, both constraint-solver arms, on RTX 5090, RTX 5070, L40S, A100-PCIE-40GB and A100-SXM4-80GB, with genesis-world 1.3.3 and `main` (01387ec8). `cuda-gdb` on the frozen process shows one resident kernel (`_func_narrowphase_multicontact…` or `_func_narrowphase_contact0`), one active lane inside `mpr_refine_portal` reached through `func_mpr_contact`, its 31 warp-mates parked. With the loop bounded, the same trajectory records a cap exactly in the window where the unbounded build stops: A100, seed 1 — stock stops at PPO update 49, the cap fires at per-env step 6384 inside the (6272, 6400] window, the run completes 190 updates with 7 caps; seed 2 — stock stops at update 5, first cap at step 656, 20 caps in 190 updates; RTX 5090 — same picture on both arms.

A second uncapped loop of the same class: the three `while is_improving*` chases in `multi_contact.py::func_approximate_polygon_with_quad` (contact-patch path only). One NaN vertex makes every test false and the lane cycles forever.

## Steps to Reproduce

Standalone, no trainer, built on #3309's `save_checkpoint` / `load_checkpoint`. Files: https://github.com/chkxw/genesis-world/releases/tag/repro-3314 (~177 MB: `ckpt_512.gsckpt`, `ops.pkl.gz`, `meta.json`, `state_before_step512.npz`, `repro_3314_ckpt_replay.py`, `0007-pr3309-read-frame-zerocopy-fallback.patch`, `REPRO_3314.md`). A 256-env scene checkpointed right before physics step 512, plus the recorded entity calls for the following steps.

Branch as is:

```
git clone https://github.com/Genesis-Embodied-AI/genesis-world && cd genesis-world && git checkout 01387ec8
git apply 0007-pr3309-read-frame-zerocopy-fallback.patch   # #3309 checkpoint I/O with zero-copy on CUDA; reported on #3309
pip install -e .
python repro_3314_ckpt_replay.py <files dir> --ckpt-step 512 --steps 256 --timeout 120
```

Output: the restored state matches the snapshot (`max|diff|=0` for all six entities), state lines for steps 512, 528 and 544, then `HANG` — the GPU at 100 % with the lane in `mpr_refine_portal`.

With the bound:

```
git fetch https://github.com/chkxw/genesis-world fix/mpr-portal-discovery-cap && git cherry-pick a0f78f7e f018e622
python repro_3314_ckpt_replay.py <files dir> --ckpt-step 512 --steps 256 --timeout 120
```

Output: all 256 steps in about 3 s, `capped_envs=[179]` from step 560, `COMPLETED`; the printed state equals the recording's up to that portal.

Recorded on a single RTX 5090 (driver 595.84). Whether another architecture walks into the same portal from this checkpoint is untested: the collider is bit-deterministic on one card, not across cards. Resuming an exact `TrajectoryFile` (#3309) from frame 512 gives the same two outcomes.

## Expected Behavior

Both loops terminate for every input. A non-converging portal, or a NaN polygon, is dropped and reported per env, never spun on.

## Suggested fix

Fork branch `fix/mpr-portal-discovery-cap` (two commits; was #3315): cap portal discovery with `CCD_ITERATIONS` like refinement and return "no portal"; set a non-raising per-env `ErrorCode.UNCONVERGED_CONTACT_PORTAL`, exposed through `get_error_envs_mask()`, right after the attempt that hit the cap (a warm-start retry otherwise erases it); cap each quad chase at `nverts` and drop that pair's manifold. A NaN-aware comparison is not an option for the quad chase because the kernels are compiled with `NoNaNsFPMath`.

## Environment

genesis-world 1.3.3 (PyPI) and `main` 01387ec8 (+ #3309); quadrants 1.3.0; torch 2.8.0+cu128; Python 3.12; CUDA 12.8; Ubuntu 22.04 / 24.04. Drivers: 595.84 (RTX 5090), 580.173.02 (RTX 5070, L40S, A100-PCIE), 595.58.03 / 595.71.05 (A100-SXM4).

### PR DESCRIPTION
**Pull request by duburcqa**
## Description

`mpr_portal_reach_tolerance` floors the tolerance of the portal refinement with a term proportional to the projection of the new support point onto the portal normal. When the portal plane passes through the origin that projection is rounding noise, so the floor vanishes: a gap made of rounding noise counts as progress, `mpr_refine_portal` re-inserts the vertex it already holds, and the loop never ends. On the GPU one lane never leaves the narrowphase kernel and the host blocks at its next synchronisation. The floor now scales with the magnitude of the portal vertices and of the support point, which bounds the rounding error of their projections.

## Related Issue

Resolves Genesis-Embodied-AI/genesis-world#3314

Addresses Genesis-Embodied-AI/genesis-world#3310 (the same hang, first attributed to the constraint solver).

## Motivation and Context

The tolerance is relative to the smaller geom of the pair, so for a small geom touching a large one (a 1 mm box on the Go1 base against a pushed box in the report) it falls below single-precision resolution of the portal vertices, and the same dot product evaluates on both sides of zero at different sites of the loop.

## How Has This Been / Can This Be Tested?

`tests/rigid/test_collision.py::test_convex_collision_across_geom_scales` places a small box exactly touching the side face of a large yawed box at a spot where their separation sits below floating-point resolution, and checks the contact it yields. The reporter's checkpoint replay from #3314 completes with this change.

## Checklist:

- [x] I read the **CONTRIBUTING** document.
- [x] I followed the `Submitting Code Changes` section of **CONTRIBUTING** document.
- [x] I tagged the title correctly (including BUG FIX/FEATURE/MISC/BREAKING)
- [x] I updated the [documentation](https://github.com/Genesis-Embodied-AI/genesis-doc) accordingly or no change is needed.
- [x] I tested my changes and added instructions on how to test it for reviewers.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: tests/rigid/test_collision.py [hunk 1]
```diff
@@ -420,6 +420,60 @@ def test_mpr_thin_box_stack_no_lateral_phantom(show_viewer, tol):
...
+
+    face_offset = 0.5 * (BOX_SIZE + GEOM_SIZE)
+    face_normal = torch.tensor([math.cos(YAW), math.sin(YAW), 0.0], dtype=gs.tc_float, device=gs.device)
+    pos_local = torch.tensor([face_offset, *FACE_SPOT], dtype=gs.tc_float, device=gs.device)
+    geom.set_pos(gu.transform_by_quat(pos_local, box.get_quat()) + box.get_pos())
+    geom.set_quat(box.get_quat())
+
+    scene.step()
+    contacts = scene.rigid_solver.collider.get_contacts()
+    assert_allclose((contacts["normal"] @ face_normal).abs(), 1.0, tol=tol)
```
[the remark this answers is not in the document]
Comment (duburcqa) [reply tests/rigid/test_collision.py:484]: fixed
